File: G28Run.java
import java.io.*;

import static java.lang.System.*;

import java.util.*;



public class G28Run {
	public static char[] temp = new char[400];
	public static int tokenlength = 1;
	public static String code = "";
	public static String[] alltokens;
	public static String[] alloperations;
	public static String[] allvalues;
	public static Deque<Double> valuereturnfromfunction = new ArrayDeque<Double>();
	public static Deque<String> currentfunction = new ArrayDeque<String>();
	public static HashMap<String, Double> functionreturn= new HashMap<String,Double>();
	public static HashMap<String, Double> hashmapglobalvalues = new HashMap<String, Double>();
	public static HashMap<String, Boolean> booleanhashmapglobalvalues = new HashMap<String, Boolean>();
	public static HashMap<String, Stack<Integer>> stackhashmapglobalvalues = new HashMap<String, Stack<Integer>>();
	public static HashMap<String, Deque<Integer>> functions= new HashMap<String,Deque<Integer>>();
	public static Deque<Integer> functionstack = new ArrayDeque<Integer>();

	
	public static void main(String[] args) throws IOException {
	
		int n = 0 ;
		int m = 0;
		
	try{
		FileInputStream bytecode = new FileInputStream(args[0]);
		int r = bytecode.read();
		char c = (char)r;
		temp[0]= c;
		
	
		while(((r = bytecode.read()) != -1))
			{			
			c = (char)r;
			temp[tokenlength]= c;
			tokenlength += 1;
			}
		
		code = String.valueOf(temp);
		code = code.substring(0,tokenlength-1);
		alltokens = code.split("\\s+");
		alloperations = new String[tokenlength];
		allvalues  = new String[tokenlength];
		
		
			while(!alltokens[n].equals("end"))

			{ 
				if ((alltokens[n].equals("asn")) ||(alltokens[n].equals("lst")) ||(alltokens[n].equals("jbk")) ||(alltokens[n].equals("run")) ||(alltokens[n].equals("fnd")) ||(alltokens[n].equals("fun")) ||(alltokens[n].equals("add")) ||(alltokens[n].equals("sub")) ||(alltokens[n].equals("mul")) ||(alltokens[n].equals("div")) ||(alltokens[n].equals("put")) || (alltokens[n].equals("get"))|| (alltokens[n].equals("bne")) || (alltokens[n].equals("beq"))|| (alltokens[n].equals("dsp"))|| (alltokens[n].equals("grt"))|| (alltokens[n].equals("fth"))||(alltokens[n].equals("stk"))||(alltokens[n].equals("psh"))||(alltokens[n].equals("pop"))||(alltokens[n].equals("pek"))|| (alltokens[n].equals("and")) ||(alltokens[n].equals("or")) ||(alltokens[n].equals("not")))
				{alloperations[m]= alltokens[n];
				m += 1;}
				else {allvalues[m-1]= alltokens[n];

				      //j += 1;
				}
								
				n += 1;
			}
			
			alloperations[m]=alltokens[n];
			alloperations = Arrays.copyOf(alloperations,m+1);
			allvalues = Arrays.copyOf(allvalues,m+1);
	}catch (IOException e){
		out.println("ERROR1");
	}
		
		function(0,m+1);
		
	}
	
	@SuppressWarnings("null")
	public static void function(int a, int b){
		
				Deque<Double> stack = new ArrayDeque<Double>();
				Deque<Boolean> booleanstack = new ArrayDeque<Boolean>();
				Deque<Stack<Integer>> superstack = new ArrayDeque<Stack<Integer>>();
				
		
				HashMap<String, Double> hashmapvalues = new HashMap<String, Double>();
				HashMap<String, Boolean> booleanhashmapvalues = new HashMap<String, Boolean>();
				HashMap<String, Stack<Integer>>  stackhashmapvalues = new HashMap<String, Stack<Integer>>();
		
			
				String[] operations = Arrays.copyOfRange(alloperations,a,b );
				String[] values = Arrays.copyOfRange(allvalues,a,b );
				
				
				for ( int i=0; i<b-a;i++) {

					if (operations[i].equals("fth")) {   

						Scanner sc = new Scanner(System.in);
						out.println("Enter Value");
						double input = sc.nextDouble();
						stack.push(input);
					}
					
					if ((operations[i].equals("put")&&((stack.isEmpty()==false)||(booleanstack.isEmpty()==false)))) {   
						
						if((currentfunction.isEmpty())&&(stack.peek()!=null)) {
							hashmapglobalvalues.put(values[i], stack.pop());
						} else if ((currentfunction.isEmpty())&&(booleanstack.peek()!=null)) {
							booleanhashmapglobalvalues.put(values[i], booleanstack.pop());
						} else if ((!currentfunction.isEmpty())&&(booleanstack.peek()!=null)) {
							booleanhashmapvalues.put(values[i], booleanstack.pop());
						} else if(hashmapglobalvalues.get(values[i])!=null){
							hashmapglobalvalues.put(values[i],stack.pop());
						}
						else {
							hashmapvalues.put(values[i],stack.pop());
						}
					}
						
					if (operations[i].equals("stk")) {

						if(currentfunction.isEmpty()) {
							stackhashmapglobalvalues.put(values[i], new Stack<Integer>());
						}
						else{
							stackhashmapvalues.put(values[i],new Stack<Integer>());
						}
					}
					
					if(operations[i].equals("psh")) {

						if((currentfunction.isEmpty())&&(stack.peek()!=null)) {
							Stack<Integer> st = stackhashmapglobalvalues.get(values[i]);
							st.push(stack.pop().intValue());
							stackhashmapglobalvalues.put(values[i], st);
						} else {
							Stack<Integer> st = stackhashmapvalues.get(values[i]);
							st.push(stack.pop().intValue());
							stackhashmapvalues.put(values[i],st);
						}
						
					}
					
					if(operations[i].equals("pop")) {

						if(currentfunction.isEmpty()) {

							Stack<Integer> st = stackhashmapglobalvalues.get(values[i]);
							int value = st.pop();
							stack.push((double) value);
						} else {
							Stack<Integer> st = stackhashmapvalues.get(values[i]);
							int value = st.pop();
							stack.push((double) value);
						}
						
					}
					
					if(operations[i].equals("pek")) {

						if(currentfunction.isEmpty()) {
							Stack<Integer> st = stackhashmapglobalvalues.get(values[i]);
							int value = st.peek();
							stack.push((double) value);
						} else {
							Stack<Integer> st = stackhashmapvalues.get(values[i]);
							int value = st.peek();
							stack.push((double) value);
						}
						
					}
					
					if (operations[i].equals("get")) {   

						boolean doubleornot = true;

						try { 
							Double.parseDouble(values[i]); 

						} catch(NumberFormatException e) {
							doubleornot = false;
						}
						
						if(doubleornot==false) {
							
							if(hashmapglobalvalues.get(values[i])!=null) {

								stack.push(hashmapglobalvalues.get(values[i]));

							} else if(hashmapvalues.get(values[i])!=null) {

								stack.push(hashmapvalues.get(values[i]));

							} else if (functions.get(values[i])!=null) {

								int temporary = functions.get(values[i]).peek();
								functionstack.push(functions.get(values[i]).pop());
								functionstack.push(functions.get(values[i]).peek());
								functions.get(values[i]).push(temporary);

							} else if (booleanhashmapvalues.get(values[i])!=null) {

								booleanstack.push(booleanhashmapvalues.get(values[i]));

							} else if (values[i].equals("true")) {

								booleanstack.push(true);

							} else if (values[i].equals("false")) {

								booleanstack.push(false);

							} else if ((booleanhashmapglobalvalues.get(values[i])!=null)) {

								booleanstack.push(booleanhashmapglobalvalues.get(values[i]));

							} else if(stackhashmapglobalvalues.get(values[i])!=null) {

								superstack.push(stackhashmapglobalvalues.get(values[i]));

							} else if(stackhashmapvalues.get(values[i])!=null) {

								superstack.push(stackhashmapvalues.get(values[i]));
							} else {
								out.println("Error on load!");
							}
						}
						
						if(doubleornot==true) {

							stack.push(Double.parseDouble(values[i]));
						}
											
					}
					
					if ((operations[i].equals("fun"))&&(stack.isEmpty()==true)) {

						int first = i;

						while(!operations[++i].equals("fnd"));

						Deque<Integer> startandend = new ArrayDeque<Integer>();
						startandend.push(i);
						startandend.push(first+1);
						functions.put(values[first],startandend);
					}
					
					if ((operations[i].equals("jbk"))&&(stack.isEmpty()==true)) {

						boolean doubleornot = true;
						try { 
							Double.parseDouble(values[i]); 
						} catch(NumberFormatException e) {
							doubleornot = false;
						}
						
						if(doubleornot==false) {

							if(hashmapvalues.get(values[i])!=null) {

								functionreturn.put(currentfunction.peek(), hashmapvalues.get(values[i]));
							}
						}
						
						if(doubleornot==true) {

							functionreturn.put(currentfunction.peek(), Double.parseDouble(values[i]));
						}
					}
					
					if ((operations[i].equals("run"))&&(values[i]==null)) {

						int right=functionstack.pop();
						int left=functionstack.pop();
						if (left<right-1) {

							currentfunction.push(allvalues[left-1]);
							function(left,right);
							currentfunction.pop();
						}
						if (functionreturn.get(values[i-1])!=null) {

							valuereturnfromfunction.push(functionreturn.get(values[i-1]));
						}
					}
					
					if ((operations[i].equals("add"))&&(stack.isEmpty()==false)) {

						double right=stack.pop();
						double left=stack.pop();
						stack.push(left+right);
					}
					
					if ((operations[i].equals("sub"))&&(stack.isEmpty()==false)) {

						double right=stack.pop();
						double left=stack.pop();
						stack.push(left-right);
					}
					
					if ((operations[i].equals("mul"))&&(stack.isEmpty()==false)) {

						double right=stack.pop();
						double left=stack.pop();
						stack.push(left*right);
					}
					
					if ((operations[i].equals("div"))&&(stack.isEmpty()==false)) {

						double right=stack.pop();
						double left=stack.pop();
						stack.push(left/right);
					}
					
					if ((operations[i].equals("grt"))&&(stack.isEmpty()==false)) {

						double right=stack.pop();
						double left=stack.pop();

						if (left>right) {
							stack.push(1.0);
						} else {
							stack.push(0.0);
						}
					}
					
					if ((operations[i].equals("lst"))&&(stack.isEmpty()==false)) {

						double right=stack.pop();
						double left=stack.pop();

						if (left<right) {
							stack.push(1.0);
						} else {
							stack.push(0.0);
						}
					}
					if ((operations[i].equals("asn"))&&(stack.isEmpty()==false)) {

						double right=stack.pop();
						double left=stack.pop();

						if (left==right) {
							stack.push(1.0);
						} else { 
							stack.push(0.0);
						}
					}
					if ((operations[i].equals("asn"))&&(booleanstack.isEmpty()==false)) {

						Boolean right = booleanstack.pop();
						Boolean left  = booleanstack.pop();

						if (left==right) {
							stack.push(1.0);
						} else {
							stack.push(0.0);
						}
					}

					if ((operations[i].equals("and"))&&(booleanstack.isEmpty()==false)) {

						Boolean right = booleanstack.pop();
						Boolean left  = booleanstack.pop();

						if (left && right) {
							booleanstack.push(true);
						} else {
							booleanstack.push(false);
						}
					}

					if ((operations[i].equals("or"))&&(booleanstack.isEmpty()==false)) {

						Boolean right = booleanstack.pop();
						Boolean left  = booleanstack.pop();

						if (left || right) {
							booleanstack.push(true);
						} else {
							booleanstack.push(false);
						}
					}

					if ((operations[i].equals("not"))&&(booleanstack.isEmpty()==false)) {

						Boolean value = booleanstack.pop();
						booleanhashmapglobalvalues.put(values[i-1], !value);

						if (!value) {
							booleanstack.push(true);
						} else {
							booleanstack.push(false);
						}
					}

					if ((operations[i].equals("bne"))&&(stack.isEmpty()==false)) {

						if((stack.peek()==0.0)&&(currentfunction.isEmpty())) {
							i = Integer.parseInt(values[i])-1;
							stack.pop();
						} else if((stack.peek()==0.0)&&(!currentfunction.isEmpty())) {
							i = Integer.parseInt(values[i])-a;
							stack.pop();

							if(i>b-a-1) {
								i=b-a-1;
							}

						} else {
							stack.pop();
						}
					}

					if ((operations[i].equals("beq"))&&(stack.isEmpty()==false)) {

						if(stack.peek()==1.0) {
							i = Integer.parseInt(values[i])-1;
							stack.pop();
						}
					}

					if (operations[i].equals("dsp")) {

						if(stack.isEmpty()==false) {

							out.println(stack.pop());

						} else if(booleanstack.isEmpty()==false) {

							out.println(booleanstack.pop());

						} else if(superstack.isEmpty()==false) {

							out.println(superstack.pop().toString());
						}
					}
				}	
			
		
	}
	

}

File: G28Run.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;
using System.Text.RegularExpressions;
using System.Threading.Tasks;

namespace G28Run
{
    class Program
    {
        #region Declarations
        public static List<char> temp = new List<char>();
        public static int tokenLength = 1;
        public static string code = string.Empty;
        public static string[] allTokens;
        public static string[] allOperations;
        public static string[] allValues;
        public static Stack<int> valueReturnFromFunction = new Stack<int>();
        public static Stack<string> currentFunction = new Stack<string>();
        public static Dictionary<string, int> functionReturn = new Dictionary<string, int>();
        public static Dictionary<string, int> intHashMapGlobalValues = new Dictionary<string, int>();
        public static Dictionary<string, bool> booleanHashmapGlobalValues = new Dictionary<string, bool>();
        public static Dictionary<string, Stack<int>> stackHashmapGlobalValues = new Dictionary<string, Stack<int>>();
        public static Dictionary<string, string> stringHashmapGlobalValues = new Dictionary<string, string>();
        public static Dictionary<string, Stack<int>> functions = new Dictionary<string, Stack<int>>();
        public static Stack<int> functionStack = new Stack<int>();
        public static Stack<int> functionArgumentStack = new Stack<int>();
        public static List<string> OpCodes = new List<string>();

        #endregion

        static void Main(string[] args)
        {
            
            if(string.IsNullOrEmpty(args[0]))
            {
                Console.WriteLine("Please input the file name and try again");
                return;
            }
            
            string fileName = args[0];
            string line = string.Empty;
            StringBuilder code = new StringBuilder();

            try
            {
                using (StreamReader file = new StreamReader(fileName))
                {
                    while((line = file.ReadLine()) != null)
                    {
                        code.Append(line + " ");
                    }
                }
                allTokens = Regex.Split(code.ToString(),@"\s+");
                allOperations = new string[allTokens.Length];
                allValues = new string[allTokens.Length];

                InitializeOpcodes();

                SeparateOperationsAndTokens();

                RunTime(0, allValues.Length);

                Console.ReadLine();
            }
            catch(Exception ex)
            {
                Console.WriteLine("Error found " + ex.Message);
            }
        }

        public static void RunTime(int start, int end)
        {
            // Declarations
            Stack<int> intstack = new Stack<int>();
            Stack<bool> boolStack = new Stack<bool>();
            Stack<string> stringStack = new Stack<string>();
            Stack<Stack<int>> superStack = new Stack<Stack<int>>();

            Dictionary<string, int> intHashMap = new Dictionary<string, int>();
            Dictionary<string, bool> boolHashMap = new Dictionary<string, bool>();
            Dictionary<string, string> stringHashMap = new Dictionary<string, string>();
            Dictionary<string, Stack<int>> stackHashMap = new Dictionary<string, Stack<int>>();

            string opCode, operand;
            string[] operations = new string[(end - start)];
            string[] values = new string[(end - start)];

            Array.Copy(allOperations, start, operations, 0, (end - start));
            Array.Copy(allValues, start, values, 0, (end - start));

            for(int i = 0; i < end - start; i++)
            {
                opCode = operations[i];
                operand = values[i];

                if (opCode.Equals("fth"))
                {
                    Console.WriteLine("Enter the value: ");
                    int input = Int32.Parse(Console.ReadLine());
                    intstack.Push(input);
                }

                else if(opCode.Equals("put") && ((intstack.Count > 0) || (boolStack.Count > 0) || (valueReturnFromFunction.Count > 0) || (stringStack.Count > 0)))
                {
                    if(currentFunction.Count == 0 && intstack.Count > 0)
                    {
                        intHashMapGlobalValues[operand] = intstack.Pop();
                    }
                    else if(currentFunction.Count == 0 && boolStack.Count > 0)
                    {
                        booleanHashmapGlobalValues[operand] = boolStack.Pop();
                    }
                    else if(currentFunction.Count > 0 && intstack.Count > 0)
                    {
                        if(intHashMap.ContainsKey(operand))
                        {
                            intHashMap[operand] = intstack.Pop();
                        }
                        else if(intHashMapGlobalValues.ContainsKey(operand))
                        {
                            intHashMapGlobalValues[operand] = intstack.Pop();
                        }
                        else
                        {
                            intHashMap[operand] = intstack.Pop();
                        }
                    }
                    else if(currentFunction.Count > 0 && boolStack.Count > 0)
                    {
                        boolHashMap[operand] = boolStack.Pop();
                    }
                    else if(currentFunction.Count == 0 && stringStack.Count > 0)
                    {
                        stringHashmapGlobalValues[operand] = stringStack.Pop();
                    }
                    else if(currentFunction.Count > 0 && stringStack.Count > 0)
                    {
                        stringHashMap[operand] = stringStack.Pop();
                    }
                    else if(valueReturnFromFunction.Count > 0)
                    {
                        if (intHashMap.ContainsKey(operand))
                        {
                            intHashMap[operand] = valueReturnFromFunction.Pop();
                        }
                        else if(intHashMapGlobalValues.ContainsKey(operand))
                        {
                            intHashMapGlobalValues[operand] = valueReturnFromFunction.Pop();
                        }
                        else
                        {
                            intHashMap[operand] = valueReturnFromFunction.Pop();
                        }
                    }                    
                }

                else if (opCode.Equals("str"))
                {
                    stringStack.Push(operand);
                }

                else if (opCode.Equals("stk"))
                {
                    if(currentFunction.Count == 0)
                    {
                        stackHashmapGlobalValues[operand] = new Stack<int>();
                    }
                    else
                    {
                        stackHashMap[operand] = new Stack<int>();
                    }
                }

                else if (opCode.Equals("psh"))
                {
                    if(currentFunction.Count == 0 && intstack.Count > 0)
                    {
                        Stack<int> st = stackHashmapGlobalValues[operand];
                        st.Push(intstack.Pop());
                        stackHashmapGlobalValues[operand] = st;
                    }
                    else
                    {
                        Stack<int> st = stackHashMap[operand];
                        st.Push(intstack.Pop());
                        stackHashMap[operand] = st;
                    }
                }

                else if (opCode.Equals("pop"))
                {
                    if (currentFunction.Count == 0)
                    {
                        Stack<int> st = stackHashmapGlobalValues[operand];
                        intstack.Push(st.Pop());
                    }
                    else
                    {
                        Stack<int> st = stackHashMap[operand];
                        intstack.Push(st.Pop());
                    }
                }


                else if (opCode.Equals("pek"))
                {
                    if (currentFunction.Count == 0)
                    {
                        Stack<int> st = stackHashmapGlobalValues[operand];
                        intstack.Push(st.Peek());
                    }
                    else
                    {
                        Stack<int> st = stackHashMap[operand];
                        intstack.Push(st.Peek());
                    }
                }

                else if (opCode.Equals("get"))
                {
                    bool isInteger = CheckIsInteger(operand);
                    if(!isInteger)
                    {
                        if(intHashMap.ContainsKey(operand))
                        {
                            intstack.Push(intHashMap[operand]);
                        }
                        else if (intHashMapGlobalValues.ContainsKey(operand))
                        {
                            intstack.Push(intHashMapGlobalValues[operand]);
                        }
                        else if(functions.ContainsKey(operand))
                        {
                            int temp = functions[operand].Peek();
                            functionStack.Push(functions[operand].Pop());
                            functionStack.Push(functions[operand].Peek());
                            functions[operand].Push(temp);
                        }
                        else if(boolHashMap.ContainsKey(operand))
                        {
                            boolStack.Push(boolHashMap[operand]);
                        }
                        else if(operand.Equals("true"))
                        {
                            boolStack.Push(true);
                        }
                        else if(operand.Equals("false"))
                        {
                            boolStack.Push(false);
                        }
                        else if(booleanHashmapGlobalValues.ContainsKey(operand))
                        {
                            boolStack.Push(booleanHashmapGlobalValues[operand]);
                        }
                        else if(stackHashmapGlobalValues.ContainsKey(operand))
                        {
                            superStack.Push(stackHashmapGlobalValues[operand]);
                        }
                        else if(stackHashMap.ContainsKey(operand))
                        {
                            superStack.Push(stackHashMap[operand]);
                        }
                        else if(stringHashmapGlobalValues.ContainsKey(operand))
                        {
                            stringStack.Push(stringHashmapGlobalValues[operand]);
                        }
                        else if(stringHashMap.ContainsKey(operand))
                        {
                            stringStack.Push(stringHashMap[operand]);
                        }
                        else
                        {
                            Console.WriteLine("Error on load");
                        }
                    }
                    else
                    {
                        intstack.Push(Int32.Parse(operand));
                    }
                }

                else if (opCode.Equals("gta"))
                {
                    bool isInteger = CheckIsInteger(operand);
                    if (!isInteger)
                    {
                        if (intHashMap.ContainsKey(operand))
                        {
                            functionArgumentStack.Push(intHashMap[operand]);
                        }
                        else if (intHashMapGlobalValues.ContainsKey(operand))
                        {
                            functionArgumentStack.Push(intHashMapGlobalValues[operand]);
                        }
                    }
                    else
                    {
                        functionArgumentStack.Push(Int32.Parse(operand));
                    }
                }

                else if (opCode.Equals("pta"))
                {
                    if (functionArgumentStack.Count > 0)
                    {
                        intHashMap[operand] = functionArgumentStack.Pop();
                    }
                }

                else if (opCode.Equals("fun") && intstack.Count == 0)
                {
                    int temp = i;
                    while(!operations[++i].Equals("fnd"))
                    {
                       ;
                    }
                    Stack<int> startAndEnd = new Stack<int>();
                    startAndEnd.Push(i);
                    startAndEnd.Push(temp + 1);
                    functions.Add(operand, startAndEnd);
                }

                else if (opCode.Equals("jbk") && intstack.Count == 0)
                {
                    bool isInteger = CheckIsInteger(operand);
                    if(!isInteger)
                    {
                        if(intHashMap.ContainsKey(operand))
                        {
                            functionReturn[currentFunction.Peek()] = intHashMap[operand];
                        }
                        else if(intHashMapGlobalValues.ContainsKey(operand))
                        {
                            functionReturn[currentFunction.Peek()] = intHashMapGlobalValues[operand];
                        }
                    }
                    else
                    {
                        functionReturn[currentFunction.Peek()] = Int32.Parse(operand);
                    }
                    i = 100;
                }

                else if (opCode.Equals("run") && (operand == null || operand == string.Empty))
                {
                    int right = functionStack.Pop();
                    int left = functionStack.Pop();
                    if(left < right)
                    {
                        currentFunction.Push(allValues[left-1]);
                        RunTime(left, right);
                        currentFunction.Pop();
                    }
                    if(functionReturn.ContainsKey(values[i - 1]))
                    {
                        valueReturnFromFunction.Push(functionReturn[values[i - 1]]);
                        //intstack.Push(functionReturn[values[i - 1]]);
                    }
                }


                else if ((intstack.Count > 0) && (opCode.Equals("add") || opCode.Equals("sub") || opCode.Equals("mul") || opCode.Equals("div") ||
                    opCode.Equals("grt") || opCode.Equals("lst") || opCode.Equals("asn")))
                {
                    bool isReturnValue = false;
                    int right = intstack.Pop();
                    int left;
                    if (intstack.Count > 0)
                    {
                        left = intstack.Pop();
                    }
                    else
                    {
                        isReturnValue = true;
                        left = valueReturnFromFunction.Pop();
                    }

                    switch(opCode)
                    {
                        case "add":
                                    if (!isReturnValue)
                                    {
                                        intstack.Push(left + right);
                                    }
                                    else
                                    {
                                        valueReturnFromFunction.Push(left + right);
                                    }
                                    break;
                        case "sub":
                                    if (!isReturnValue)
                                    {
                                        intstack.Push(left - right);
                                    }
                                    else
                                    {
                                        valueReturnFromFunction.Push(left - right);
                                    }
                                    break;
                        case "mul": 
                                    if (!isReturnValue)
                                    {
                                        intstack.Push(left * right);
                                    }
                                    else
                                    {
                                        valueReturnFromFunction.Push(left * right);
                                    }
                                    break;
                        case "div":
                                    if (!isReturnValue)
                                    {
                                        intstack.Push(left / right);
                                    }
                                    else
                                    {
                                        valueReturnFromFunction.Push(left / right);
                                    }
                                    break;
                        case "grt":
                                    if(left > right)
                                    {
                                        intstack.Push(1);
                                    }
                                    else
                                    {
                                        intstack.Push(0);
                                    }
                                    break;
                        case "lst":
                                    if (left < right)
                                    {
                                        intstack.Push(1);
                                    }
                                    else
                                    {
                                        intstack.Push(0);
                                    }
                                    break;

                        case "asn":
                                    if (left == right)
                                    {
                                        intstack.Push(1);
                                    }
                                    else
                                    {
                                        intstack.Push(0);
                                    }
                                    break;
                    }
                }

                else if ((boolStack.Count > 0) && (opCode.Equals("asn") || opCode.Equals("and") || opCode.Equals("or")))
                {
                    bool right = boolStack.Pop();
                    bool left = boolStack.Pop();

                    switch (opCode)
                    {
                        case "asn":
                            if (left == right)
                            {
                                boolStack.Push(true);
                            }
                            else
                            {
                                boolStack.Push(false);
                            }
                            break;

                        case "and":
                            if (left && right)
                            {
                                boolStack.Push(true);
                            }
                            else
                            {
                                boolStack.Push(false);
                            }
                            break;
                        case "or":
                            if (left || right)
                            {
                                boolStack.Push(true);
                            }
                            else
                            {
                                boolStack.Push(false);
                            }
                            break;
                    }
                }

                else if (boolStack.Count > 0 && opCode.Equals("not"))
                {
                    bool value = boolStack.Pop();
                    if (!value)
                    {
                        boolStack.Push(true);
                    }
                    else
                    {
                        boolStack.Push(false);
                    }
                    break;
                }

                else if (opCode.Equals("bne") && intstack.Count > 0)
                {
                    if(intstack.Peek() == 0)
                    {
                        if (currentFunction.Count == 0)
                        {
                            i = Int32.Parse(values[i]) - 1;
                            intstack.Pop();
                        }
                        else
                        {
                            i = Int32.Parse(values[i]) - start;
                            intstack.Pop();
                            if(i > end - start - 1)
                            {
                                i = end - start - 1;
                            }
                            else
                            {
                                i--;
                            }
                        }
                    }
                    else
                    {
                        intstack.Pop();
                    }
                }

                else if (opCode.Equals("beq") && intstack.Count > 0)
                {
                    if (intstack.Peek() == 1)
                    {
                        i = Int32.Parse(values[i]) - 1;
                        intstack.Pop();
                    }
                }

                else if (opCode.Equals("dsp"))
                {
                    if(intstack.Count > 0)
                    {
                        Console.WriteLine(intstack.Pop());
                    }
                    else if(boolStack.Count > 0)
                    {
                        Console.WriteLine(boolStack.Pop());
                    }
                    else if(superStack.Count > 0)
                    {
                        foreach (int ele in superStack.Pop())
                        {
                            Console.WriteLine(ele + " ");
                        }
                    }
                    else if(stringStack.Count > 0)
                    {
                        Console.WriteLine(stringStack.Pop().Trim('"')); 
                    }
                }

            }
        }

        public static bool CheckIsInteger(string operand)
        {
            bool isInteger = true;
            try
            {
                Int32.Parse(operand);
            }
            catch (Exception ex)
            {
                isInteger = false;
            }
            return isInteger;
        }

        public static void SeparateOperationsAndTokens()
        {
            int m = 0;
            foreach(string s in allTokens)
            {
                if(!s.Equals("end"))
                {
                    if (OpCodes.Contains(s))
                    {
                        allOperations[m++] = s; 
                    }
                    else
                    {
                        allValues[m-1] = s; 
                    }
                }
            }
            allOperations[m] = allTokens.Last();
            Array.Resize(ref allOperations, m+1);
            Array.Resize(ref allValues, m+1);

        }

        public static void InitializeOpcodes()
        {
            string[] list = new string[] { "asn", "lst", "jbk", "run", "fnd", "fun", "add", "sub", "mul", "div", "put", "get",
            "bne", "beq", "dsp", "grt", "fth", "and", "or", "not","psh", "pop", "pek", "stk", "gta", "pta", "str"};
            foreach(string s in list)
            {
                OpCodes.Add(s);
            }
        }
    }
}
